com.unity.ml-agents/Tests/Editor/Sensor/StackingSensorTests.cs

```diff
@@ -60,5 +63,147 @@ public void TestStackingReset()
+        {
+            public SensorCompressionType CompressionType = SensorCompressionType.PNG;
+            public int[] Mapping;
+            public int[] Shape;
+            public float[,,] CurrentObservation;
+
+            internal Dummy3DSensor()
+            {
+            }
+
+            public int[] GetObservationShape()
+            {
+                return Shape;
+            }
+
+            public int Write(ObservationWriter writer)
+            {
+                for (var h = 0; h < Shape[0]; h++)
+                {
+                    for (var w = 0; w < Shape[1]; w++)
+                    {
+                        for (var c = 0; c < Shape[2]; c++)
+                        {
+                            writer[h, w, c] = CurrentObservation[h, w, c];
+                        }
+                    }
+                }
+                return Shape[0] * Shape[1] * Shape[2];
+            }
+
+            public byte[] GetCompressedObservation()
+            {
+                var writer = new ObservationWriter();
+                var flattenedObservation = new float[Shape[0] * Shape[1] * Shape[2]];
+                writer.SetTarget(flattenedObservation, Shape, 0);
+                Write(writer);
+                byte[] bytes = Array.ConvertAll(flattenedObservation, (z) => (byte)z);
+                return bytes;
+            }
+
+            public void Update() { }
+
+            public void Reset() { }
+
+            public SensorCompressionType GetCompressionType()
+            {
+                return CompressionType;
+            }
+
+            public string GetName()
+            {
+                return "Dummy";
+            }
+
+            public int[] GetCompressedChannelMapping()
+            {
+                return Mapping;
+            }
+        }
+
+        [Test]
+        public void TestStackingMapping()
+        {
+            // Test grayscale stacked mapping with CameraSensor
+            var cameraSensor = new CameraSensor(new Camera(), 64, 64,
+                true, "grayscaleCamera", SensorCompressionType.PNG);
+            var stackedCameraSensor = new StackingSensor(cameraSensor, 2);
+            Assert.AreEqual(stackedCameraSensor.GetCompressedChannelMapping(), new[] { 0, 0, 0, 1, 1, 1 });
+
+            // Test RGB stacked mapping with RenderTextureSensor
+            var renderTextureSensor = new RenderTextureSensor(new RenderTexture(24, 16, 0),
+                false, "renderTexture", SensorCompressionType.PNG);
+            var stackedRenderTextureSensor = new StackingSensor(renderTextureSensor, 2);
+            Assert.AreEqual(stackedRenderTextureSensor.GetCompressedChannelMapping(), new[] { 0, 1, 2, 3, 4, 5 });
+
+            // Test mapping with dummy layers that should be dropped
+            var dummySensor = new Dummy3DSensor();
+            dummySensor.Shape = new int[] { 2, 2, 4 };
+            dummySensor.Mapping = new int[] { 0, 1, 2, 3, -1, -1 };
```

Comment:
The wrapped sensor shouldn't be adding -1's to its mapping, right? I think we decided this was the responsibility of the StackingSensor. The assert below can stay the same, though.

Reply:
Actually, I guess you should try both - once where -1 is in the wrapped sensor's mapping to make sure they're preserved, and another to make sure they're added if needed.

Reply:
yes I agree both cases can happen, added test for that.